drivers/spi/spi_andes_atcspi200.c

```diff
@@ -162,31 +191,414 @@ static int configure(const struct device *dev,
+
+static inline void spi_tx_dma_enable(const struct device *dev)
 {
-	struct spi_atcspi200_data * const data = dev->data;
+	const struct spi_atcspi200_cfg * const cfg = dev->config;
+	/* Enable TX DMA */
+	sys_set_bits(SPI_CTRL(cfg->base), CTRL_TX_DMA_EN_MSK);
+}
+
+static inline void spi_tx_dma_disable(const struct device *dev)
+{
+	const struct spi_atcspi200_cfg * const cfg = dev->config;
+	/* Disable TX DMA */
+	sys_clear_bits(SPI_CTRL(cfg->base), CTRL_TX_DMA_EN_MSK);
+}
+
+static inline void spi_rx_dma_enable(const struct device *dev)
+{
+	const struct spi_atcspi200_cfg * const cfg = dev->config;
+	/* Enable RX DMA */
+	sys_set_bits(SPI_CTRL(cfg->base), CTRL_RX_DMA_EN_MSK);
+}
+
+static inline void spi_rx_dma_disable(const struct device *dev)
+{
+	const struct spi_atcspi200_cfg * const cfg = dev->config;
+	/* Disable RX DMA */
+	sys_clear_bits(SPI_CTRL(cfg->base), CTRL_RX_DMA_EN_MSK);
+}
+
+static int spi_dma_move_buffers(const struct device *dev)
+{
+	struct spi_atcspi200_data *data = dev->data;
 	struct spi_context *ctx = &data->ctx;
-	int error = 0;
+	uint32_t error = 0;
 
-	size_t chunk_len = spi_context_max_continuous_chunk(ctx);
+	data->dma_rx.dma_blk_cfg.next_block = NULL;
+	data->dma_tx.dma_blk_cfg.next_block = NULL;
 
-	if (chunk_len > 0) {
-		data->chunk_len = chunk_len;
-		error = spi_transfer(dev, chunk_len);
-		if (error == 0) {
+	if (spi_context_tx_on(ctx)) {
+		error = spi_dma_tx_load(dev);
+		if (error != 0) {
+			return error;
+		}
+	}
+
+	if (spi_context_rx_on(ctx)) {
+		error = spi_dma_rx_load(dev);
+		if (error != 0) {
+			return error;
+		}
+	}
+
+	return 0;
+}
+
+static inline void dma_rx_callback(const struct device *dev, void *user_data,
+				uint32_t channel, int status)
+{
+	struct spi_atcspi200_data *data = user_data;
+	const struct device *spi_dev = data->self_dev;
+	struct spi_context *ctx = &data->ctx;
+	int error;
+
+	dma_stop(data->dma_rx.dma_dev, data->dma_rx.channel);
+	spi_rx_dma_disable(spi_dev);
+
+	if (spi_context_rx_on(ctx)) {
+		if (spi_dma_rx_load(spi_dev) != 0) {
+			return;
+		}
+		spi_rx_dma_enable(spi_dev);
+		error = dma_start(data->dma_rx.dma_dev, data->dma_rx.channel);
+		__ASSERT(error == 0, "dma_start was failed in rx callback");
+	}
+}
+
+static inline void dma_tx_callback(const struct device *dev, void *user_data,
+				uint32_t channel, int status)
+{
+	struct spi_atcspi200_data *data = user_data;
+	const struct device *spi_dev = data->self_dev;
+	struct spi_context *ctx = &data->ctx;
+	int error;
+
+	dma_stop(data->dma_tx.dma_dev, data->dma_tx.channel);
+	spi_tx_dma_disable(spi_dev);
+
+	if (spi_context_tx_on(ctx)) {
+		if (spi_dma_tx_load(spi_dev) != 0) {
 			return;
 		}
+		spi_tx_dma_enable(spi_dev);
+		error = dma_start(data->dma_tx.dma_dev, data->dma_tx.channel);
+		__ASSERT(error == 0, "dma_start was failed in tx callback");
+	}
+}
+
+/*
+ * dummy value used for transferring NOP when tx buf is null
+ * and use as dummy sink for when rx buf is null
+ */
+uint32_t dummy_rx_tx_buffer;
```

Comment:
will this work for multi instances? can we just move this into device data / runtime stack?

Reply:
Yes, it work for multi instances, the data send/to this variable is noused.
I think move into runtime stack may cause some problem, because after the spi_dma_tx/rx_load function has return, the variable will be release, after that the address may be used by other function. But at that time, the transfer may still ongoing, so the variable used by other function in the address will be polluted.